### PR TITLE
Ports "Ship construction through the power of science!" and others

### PR DESCRIPTION
**Pull request by nmajask**
# Github documenting your Pull Request

This PR ports 3 PRs from Beestation

- [Ship construction through the power of science! #1398](https://github.com/BeeStation/BeeStation-Hornet/pull/1398)
- [Revamped shuttle creation through the power of science! #1904](https://github.com/BeeStation/BeeStation-Hornet/pull/1904)
- [Allows custom shuttles to land on some forms of plating, but not the … #1529](https://github.com/BeeStation/BeeStation-Hornet/pull/1529)

All of these are done by the wonderful [PowerfulBacon](https://github.com/PowerfulBacon) and most of the changes were to make it work with our code. There are still a few issues that need to be fixed, but you can still build and fly a shuttle.

## Whats Different?

The big difference is the icon for the RSD and its interface, using icons I threw together as the current ones on Bee don't fit our art style and the old one doesn't fit its function (it originally worked like an RCD, but now functions as a drone controller). These icons aren't the best, and if someone provides better ones I can replace them with it

New Icon (its animated in game):
RSD

## Whats Needs to be Fixed?
I haven't done much testing, but these are a few of the issues I have noticed.

- [x] Clean up the code
- [x] Rearrange the tech web to fit the new tech
- [x] Fix the RSD drone not using its icons
- [x] Readd the name filter
- [x] Port "[Bluespace Navigation Gigabeacon #44807](https://github.com/tgstation/tgstation/pull/44807)" by [Dennok](https://github.com/Dennok), as it allows you to make custom docks for ships

After further testing, I found a few more issues

- [x] Runtime in shuttle_engine.dm,121: Division by zero
- [x] Runtime in shuttle_creator_console.dm,61: Cannot read null.client (cause of the eye icon issue?)

Jamie informed me of some other fixes that I will look into adding when I can

- [x] "[Custom shuttle offsets and view range depends on ship size. #2685](https://github.com/BeeStation/BeeStation-Hornet/pull/2685)" by [Dennok](https://github.com/Dennok)
- [x] "[Custom transit dir for custom shuttle #2689](https://github.com/BeeStation/BeeStation-Hornet/pull/2689)" by [Dennok](https://github.com/Dennok)

There are still a ton of things I need to test, but as it is time consuming to make shuttles and test them it may take some time

## What Do You Plan on Adding?
There are a few things that I want to try adding, but most of these are just ideas and probably wont be added.

- Tgui for the shuttle interface
- More engines, like ones that have different properties depending on the gas mix provided, big engines, or one that runs on dilithium
- More upgrades (fuel usage reduction, anti-knockdown, DNA locks, ect.)
- Emaging the computer to allow you to fly to syndicate lab on Lavaland (if it is spawned)
- Ship weapons (I already looked into adding the weapons from "[Making shuttles great(er) again through the power of science! (Bluespace Exploration Core) #2551](https://github.com/BeeStation/BeeStation-Hornet/pull/2551)", but I would need more time to port it)
- Maybe even FTL-like exploration instead of our current space exploration system (This would take a LONG time to make, would only be a drain on the server if there is no job that uses it, and may not even be fun, but you could do a lot of cool stuff with it. PowerfulBacon has already created something similar too.)
- The ability to make multi area shuttles (this wouldn't be too hard in theory, as shuttles already store its contained areas as a list, so I would just need to add a way to make different areas in the RSD)

# Wiki Documentation

If this gets merged, there are going to be a lot of changes needed. The new tool, boards, machines, and techs will need to be added into their respective pages and many new icons will need to be added. Also, it may be a good idea to make a guide on how to make shuttles at some point, maybe merging it with [the one for space pods](https://wiki.yogstation.net/wiki/Space_Pods) and renaming it to space vehicle construction or something.

# Changelog

:cl:  PowerfulBacon, Dennok
rscadd: Added custom shuttles
rscadd: Added in new circuit boards, machines, techs, icons, and a tool for creating custom shuttles
rscadd: Created 2 new config options defining shuttle max size and count of shuttles
/:cl:

### CODE REVIEW

**Comment by Anvilman6** [comment]
I look at this, and I feel joy that we have such an ambitious featurecoder contributing to our base. 
But I also feel fear about how much documentation this is going to need generated.

Your PRs and ports continue to amaze, Nmajask - keep it up!

**Comment by maxion12345** [comment]
fuckkk we could have this be some sorta progression

we make it so lavaland only has pitiful amounts of diamond/bluespace, and science needs to construct a shuttle for the miners to go to space to miner asteroids that have larger amounts of them

but this will never occur because people already bitch about science having to push a button to get mining tech, having them build a shuttle would make shaft miner mains cry

**Review by nesisknesis — APPROVED** [review]

**Comment by nmajask** [comment]
Just fixed the name filter and am currently working on changing the UI for shuttle to use Tgui.

> I look at this, and I feel joy that we have such an ambitious featurecoder contributing to our base.
> But I also feel fear about how much documentation this is going to need generated.
> 
> Your PRs and ports continue to amaze, Nmajask - keep it up!

Thanks for the support, I have many more things that I want to add and port. Also, if you need any help with updating the wiki with changes from my PRs I would be more than happy to help provide information or make edits/guides.

> fuckkk we could have this be some sorta progression
> 
> we make it so lavaland only has pitiful amounts of diamond/bluespace, and science needs to construct a shuttle for the miners to go to space to miner asteroids that have larger amounts of them
> 
> but this will never occur because people already bitch about science having to push a button to get mining tech, having them build a shuttle would make shaft miner mains cry

Yeah, it would be cool to have shuttles be used for progression in mining, but what would more likely happen is miners just wouldn't care and sci would just end up with less resources to work with. The only way you could get that to work is if you either rework mining and made it less focused on fighting bosses and getting loot or add bosses and loot to asteroids.

**Comment by PowerfulBacon** [comment]
Should be noted that the engines and heaters cause hard dels due to not doing anything when deleted.

**Comment by nmajask** [comment]
I have been holding off working on this for WAY too long

File: code/modules/shuttle/custom_shuttle.dm [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+/obj/machinery/computer/camera_advanced/shuttle_docker/custom/attack_hand(mob/user)
+	if(!shuttleId)
+		to_chat(user, "<span class='warning'>You must link the console to a shuttle first.</span>")
+		return
+	return ..()
+
+/obj/machinery/computer/camera_advanced/shuttle_docker/custom/proc/linkShuttle(var/new_id)
+	shuttleId = new_id
+	shuttlePortId = "shuttle[new_id]_custom"
```
Comment (PowerfulBacon) [inline code/modules/shuttle/custom_shuttle.dm:265]: This used to break on bee if you make more than 1 shuttle docker so might want to check it out

Comment (nmajask) [reply code/modules/shuttle/custom_shuttle.dm:268]: > Should be noted that the engines and heaters cause hard dels due to not doing anything when deleted.

Sorry for taking so long to reply, I got burnt out from working on this for a while and just now got back to working on it. I talked with Jamie about this and he said that its not something to worry about.

**Comment by nmajask** [comment]
> Should be noted that the engines and heaters cause hard dels due to not doing anything when deleted.

Sorry for taking so long to reply, I got burnt out from working on this for a while and just now got back to working on it. I talked with Jamie about this and he said that its not something to worry about.

File: code/modules/shuttle/custom_shuttle.dm (hunk 1 above)
Comment (nmajask) [reply code/modules/shuttle/custom_shuttle.dm:265]: > This used to break on bee if you make more than 1 shuttle docker so might want to check it out

Was there a PR that fixed it, or did you instead just rework the shuttle system to get around it?

**Comment by SomeguyManperson** [comment]
is this ready for testing?

File: code/modules/shuttle/custom_shuttle.dm (hunk 1 above)
Comment (PowerfulBacon) [reply code/modules/shuttle/custom_shuttle.dm:265]: It was fixed as a part of the shuttle system rework

Comment (PowerfulBacon) [reply code/modules/shuttle/custom_shuttle.dm:265]: I think the cause was the port ID not being unique for each console. Just adding a static number and sticking _[varname++] on the end should fix it I think. (Actually the shuttles themselves would need to have the custom port added too so maybe its not as easy as it sounds)

Comment (PowerfulBacon) [reply code/modules/shuttle/custom_shuttle.dm:265]: It was either that or the ports weren't being deleted properly.

**Comment by nmajask** [comment]
> is this ready for testing?

It should be, I got everything that was vital for it to work done I think

**Comment by JamieD1** [comment]
Test merging

**Review by SomeguyManperson — APPROVED** [review]
currently has a merge conflict I don't think we've got any bugs or massive issues so

**Comment by JamieD1** [comment]
Fix conflict then dm me

**Pull request merged**